stan/math/rev/fun/columns_dot_product.hpp

```diff
@@ -19,12 +20,49 @@ template <typename Mat1, typename Mat2,
           require_any_eigen_vt<is_var, Mat1, Mat2>* = nullptr>
 inline Eigen::Matrix<return_type_t<Mat1, Mat2>, 1, Mat1::ColsAtCompileTime>
 columns_dot_product(const Mat1& v1, const Mat2& v2) {
-  check_matching_sizes("dot_product", "v1", v1, "v2", v2);
-  Eigen::Matrix<var, 1, Mat1::ColsAtCompileTime> ret(1, v1.cols());
-  for (size_type j = 0; j < v1.cols(); ++j) {
-    ret.coeffRef(j) = dot_product(v1.col(j), v2.col(j));
+  check_matching_sizes("columns_dot_product", "v1", v1, "v2", v2);
+
+  ref_type_t<Mat1> v1_ref = v1;
+  ref_type_t<Mat2> v2_ref = v2;
+  
+  using Mat1_double = plain_type_t<decltype(value_of(v1_ref))>;
+  using Mat1_var = promote_scalar_t<var, Mat1_double>;
+
+  using Mat2_double = plain_type_t<decltype(value_of(v2_ref))>;
+  using Mat2_var = promote_scalar_t<var, Mat2_double>;
+
+  arena_matrix<Mat1_double> arena_v1_val = value_of(v1_ref);
```

Comment:
I'm still not totally happy with these value_ofs. There will be one unnecessary copy if one input is autodiff and one is double.

Reply:
I don't want to depend on value_of being an expression though. If someone changed the definition of value_of to be not-an-expression this would break.

Reply:
That is a good point. Also it makes `to_arena_if` useless as it is implemented now. I think we can change it so it returns something that never allocates dynmaic memory, for example an empty `Map` if the condition is false. I will open a PR with this.